### PR TITLE
Stdlib: add Fun module.

### PR DESCRIPTION
**Pull request by dbuenzli**
This PR adds a `Fun` module to the standard library. This will prevent the new `protect` function added by #1855 and being refined in #2118 to be added to the initial scope (aka `Stdlib`, ex-`Pervasives`).

A few notes:

1. The PR provides a few functions that can be systematically found in containers' [`CCFun`](https://github.com/c-cube/ocaml-containers/blob/master/src/core/CCFun.mli) module, base's [`Fn`](https://github.com/janestreet/base/blob/master/src/fn.mli) module and the Haskell [Prelude](https://hackage.haskell.org/package/base-4.12.0.0/docs/Prelude.html). These are exactly: `const`, `id` and `flip`.

2. While preparing this PR I discovered that the function `Bool.negate` introduced by PR #2010 actually lives in `Base` under the name `Fn.non`. We could consider moving `Bool.negate` to `Fun.negate`, it seems to me the latter would might make more sense from a reading point of view. If there is agreement I can add a commit to this PR to implement this.

### CODE REVIEW

File: manual/manual/library/stdlib.etex [hunk 1]
```diff
@@ -99,6 +99,10 @@ be called from C \\
 "Spacetime" & p.~\pageref{Spacetime} & memory profiler \\
 "Sys" & p.~\pageref{Sys} & system interface \\
 \end{tabular}
+\subsubsection{Misc:}
+\begin{tabular}{lll}
+"Fun" & p.~\pageref{Fun} & Function values \\
```
Comment (alainfrisch) [inline manual/manual/library/stdlib.etex]:
```suggestion
"Fun" & p.~\pageref{Fun} & function values \\
```
(or functional)

**Review by alainfrisch — APPROVED** [review]
LGTM.  (The link to Base's `Fn`  should be https://github.com/janestreet/base/blob/master/src/fn.mli )

To be clear, do you intend to move `protect` in a later PR, or did you intend the move to be part of this one but forgot about it?

(I suggest to discuss more additions to `Fun` in subsequent PRs.)

File: manual/manual/library/stdlib.etex (hunk 1 above)
Comment (dbuenzli) [reply manual/manual/library/stdlib.etex]: It seems there's a mix of capital vs non-capitalized entries there but I can do this. 

Can a native english speaker please comment on "functional" vs "function" values, I hesitated quite a bit and the www didn't provide any answer.

**Comment by dbuenzli** [comment]
> To be clear, do you intend to move `protect` in a later PR, 

Yes or if this is gets merged before the work on #2118 ends. I expect the latter to do the move directly.

What do you think about point 2. ?

**Comment by alainfrisch** [comment]
Re 2: yes, Fun.negate reads better than Bool.negate.

**Comment by alainfrisch** [comment]
(In terms of future addition, let me suggest `('a -> 'b) -> ('a -> ('b, exn) Result.t)` -- which I would have called `protect` if the name was not already taken.)

**Comment by dbuenzli** [comment]
> (In terms of future addition, let me suggest `('a -> 'b) -> ('a -> ('b, exn) Result.t)` -- which I would have called `protect` if the name was not already taken.)

I have these things in `Rresult` [here](http://erratique.ch/software/rresult/doc/Rresult.R.html#exn) under the name `trap_exn`, but I'm not sure they are a good idea: they imply a catch all exception handler but some exceptions like `Break` should not be caught (maybe we can list those though).

**Comment by alainfrisch** [comment]
>  but some exceptions like Break should not be caught

This is a very good reason to expose such a function since it's unlikely that user-defined version would be fully correct.  Or perhaps a predicate `Exn.catchable: exn -> bool` (or whatever).

File: manual/manual/library/stdlib.etex (hunk 1 above)
Comment (pmetzger) [reply manual/manual/library/stdlib.etex]: "Function" feels better here. "Functional" probably would confuse some people, though it wouldn't be wrong as such.

**Comment by dbuenzli** [comment]
Thanks !

**Pull request merged**